src/libopensc/pkcs15-pin.c

```diff
@@ -306,8 +306,11 @@ sc_pkcs15_verify_pin(struct sc_pkcs15_card *p15card, struct sc_pkcs15_object *pi
 	 * in this case, to avoid error or unnecessary pin prompting on pinpad,
 	 * check if the PIN has been already verified and the access condition
 	 * is still open on card.
+	 *
+	 * But if context specific login we need to use pin pad reader
+	 * and not cache the pin 
 	 */
-	if (pinlen == 0) {
+	if (pinlen == 0 && auth_info->auth_method != SC_AC_CONTEXT_SPECIFIC) {
```

Comment:
Just the opposite. It fixes a problem with using a PIN PAD reader with  CKA_ALWAYS_AUTHENTICATE. Without the change the code skips doing the PIN PAD verify. Read the comments in the code above the patch  lines 304-308, and the added comments 310-311: 
```
304         /*
305          * if pin cache is disabled, we can get here with no PIN data.
306          * in this case, to avoid error or unnecessary pin prompting on pinpad,
307          * check if the PIN has been already verified and the access condition
308          * is still open on card.
309          *
310          * But if context specific login we need to use pin pad reader
311          * and not cache the pin 
312          */
```
The previous code assumed if the login state of the card was logged_in, verify pin is not needed. That's true for normal verify but not a for a  CKA_ALWAYS_AUTHENTICATE/CONTEXT_SPECIFIC  verify, where the card is expecting a verify pin command before the next crypto command using the key with user_consent/CKA_ALWAYS_AUTHENTICATE.

The PIV requires the verify to be immediately before the crypto command.  OpenPGP only allows one crypto command per verify.  This might explain your  https://github.com/OpenSC/OpenSC/issues/1263#issuecomment-386853641 : " only one OpenPGP email - on the attempt to sign another OpenPGP email it fails." 

Without this patch 0.18.0 release candidates fail to work with a PIN PAD reader and a  CKA_ALWAYS_AUTHENTICATE key.